### PR TITLE
*: fix issue3356

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit fixes issue #3356 

PTAL @shenli @hanfei1991

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @zimulala

File: util/types/datum.go [hunk 1]
```diff
@@ -830,41 +830,38 @@ func (d *Datum) convertToString(sc *variable.StatementContext, target *FieldType
...
+				s = s[:tp.Flen]
+			} else if length < tp.Flen {
+				// For binary string, values shorter than specified number of bytes are padded with 0x00 bytes to the specified length.
+				padding := make([]byte, tp.Flen-length)
+				s = string(append([]byte(s), padding...))
+			}
+		} else {
+			runeArray := []rune(s)
+			if len(runeArray) > tp.Flen {
+				err = ErrDataTooLong.Gen("Data Too Long, field len %d, data len %d", tp.Flen, len(s))
```
Comment (hanfei1991) [inline util/types/datum.go]: Why not check IgnoreTruncate ?

Comment (XuHuaiyu) [inline util/types/datum.go]: sc.HandleTuncate(err) will check it in line 862.
@hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @shenli @zimulala @tiancaiamao

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@XuHuaiyu mysql-test is failed. Please fix it.

**Comment by XuHuaiyu** [comment]
pass mysql-test
PTAL @shenli

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: parser/parser.y [hunk 2]
```diff
@@ -3914,14 +3914,18 @@ CastType:
...
+		x.Collate = charset.CollationBin
 		$$ = x
 	}
 |	"CHAR" OptFieldLen OptBinary OptCharset
 	{
 		x := types.NewFieldType(mysql.TypeString)
 		x.Flen = $2.(int)
 		x.Charset = $4.(string)
+		if x.Charset == "" {
+			x.Charset = charset.CharsetUTF8
```
Comment (shenli) [inline parser/parser.y]: Consider OptBinary？

Comment (XuHuaiyu) [inline parser/parser.y]: @shenli 
I am not sure whether it's a mistake,
there is no binary option in the[ MySQL5.7 manual](https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast).

Comment (shenli) [inline parser/parser.y]: In MySQL's source code, binary option is allowed. https://github.com/mysql/mysql-server/blob/5.7/sql/sql_yacc.yy#L10190

Comment (XuHuaiyu) [inline parser/parser.y]: `OptBinary OptCharset` in TiDB is the same as `opt_binary` in MySQL.
I add the binary flag when `OptBinary` is "BINARY" refer to the MySQL source code.

Comment (XuHuaiyu) [inline parser/parser.y]: PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Pull request merged**